src/jv.c

```diff
@@ -1862,6 +1862,132 @@ jv jv_object_iter_value(jv object, int iter) {
 /*
  * Memory management
  */
+jv jv_unshare(jv input){
+	switch(jv_get_kind(input)){
+		case JV_KIND_INVALID:
+			{
+				if(!jv_invalid_has_msg(jv_copy(input))){
+					jv_free(input);
+					return jv_invalid();
+				}
+				return jv_invalid_with_msg(jv_unshare(jv_invalid_get_msg(jv_copy(input))));
+			}
+		case JV_KIND_OBJECT:
+			{
+				jv keys = jv_keys(jv_copy(input));
+				size_t keys_length = jv_array_length(jv_copy(keys));
+
+				jv output_object = jv_object();
+
+				for(size_t i = 0; i < keys_length; i++){
+					jv key = jv_array_get(jv_copy(keys), i);
+					output_object = jv_object_set(
+							output_object, jv_unshare(key),
+							jv_unshare(
+								jv_object_get(
+									jv_copy(input),
+									jv_copy(key)
+								)
+							)
+						);
+				}
+
+				jv_free(keys);
+				jv_free(input);
+				return output_object;
+			}
+		case JV_KIND_ARRAY:
+			{
+				size_t amount = jv_array_length(jv_copy(input));
+
+				jv output_array = jv_array_sized(amount);
+
+				for(size_t i = 0; i < amount; i++){
+					output_array = jv_array_set(
+							output_array,
+							i,
+							jv_unshare(
+								jv_array_get(
+									jv_copy(input),
+									i
+								)
+							)
+						);
+				}
+
+				jv_free(input);
+
+				return output_array;
+			}
+		case JV_KIND_STRING:
+			{
+				jv output_string = jv_string(jv_string_value(input));
+				jv_free(input);
+				return output_string;
+			}
+		case JV_KIND_NUMBER:
+			{
+				double val = jv_number_value(input);
+				jv_free(input);
+				return jv_number(val);
+			}
+		case JV_KIND_TRUE:
+			jv_free(input);
+			return jv_true();
+		case JV_KIND_FALSE:
+			jv_free(input);
+			return jv_false();
+		case JV_KIND_NULL:
+			jv_free(input);
+			return jv_null();
+		default:
+			return jv_invalid();
+	}
+}
+
+int jv_is_unshared(jv a){
+	if(jv_get_refcnt(a) != 1){
+		fprintf(stderr, "input refcnt != 1\n");
+		return 1;
+	}
+	if(jv_get_kind(a) == JV_KIND_OBJECT){
+		jv keys = jv_keys(jv_copy(a));
```

Comment:
I don't know much about the `jv` internals but i suspect that maybe `jv_is_unshared` is suitable to be implemented using a private new `jvp_*` helper function that uses some internal details, similar to how `jvp_object_equal` etc is implemented? then you can probably save on some allocation and refcounting.

@nicowilliams does that make sense?

Reply:
sorry for my inactivity for the past month on the thread. While having a jvp internal representation would get rid of some of the ref-counting it will add more complexity. Then again if you think it's worth I could implement `jvp_object_is_unshared()`, `jvp_array_is_unshared()` (I would refrain from adding jvp functions for the non-recursive options as I don't think there is much benefit to be had in these senarios)

Or do you think I should add string, number, true, false, null, invalid variants as well?